File: rs/src/domain/factory/follow.rs
use crate::domain::model::{follow::UserFollow, user::UserId, DateTime};

pub trait UserFollowFactory {
    fn create(&mut self, follower_id: UserId, followee_id: UserId) -> UserFollow;
}

pub struct DefaultUserFollowFactory {}

impl UserFollowFactory for DefaultUserFollowFactory {
    fn create(&mut self, follower_id: UserId, followee_id: UserId) -> UserFollow {
        UserFollow::new(follower_id, followee_id, DateTime::now())
    }
}

impl DefaultUserFollowFactory {
    pub fn new() -> Self {
        Self {}
    }
}

File: server/src/repository/follow.ts
import { UserFollow } from "../domain/model/follow";
import { ObjectID } from "../domain/model/object_id";

export interface FollowFetchOpts {
  limit: number;
  before?: Date;
}

export interface IFollowRepository {
  save(follow: UserFollow): Promise<void>;
  delete(follow: UserFollow): Promise<void>;
  getFollowers(userId: ObjectID, opts: FollowFetchOpts): Promise<UserFollow[]>;
  getFollowings(userId: ObjectID, opts: FollowFetchOpts): Promise<UserFollow[]>;
  isFollowing(followerId: ObjectID, followeeId: ObjectID): Promise<boolean>;
}
